File: packages/ui/src/components/chat/message/parts/contextToolGrouping.ts
import { isContextGroupTool, isToolPartActive, normalizeContextToolName } from './toolRenderUtils';

export type ContextToolCountKey = 'read' | 'search' | 'list';

export type ContextToolCounts = Record<ContextToolCountKey, number>;

export const CONTEXT_TOOL_COUNT_ORDER: readonly ContextToolCountKey[] = ['search', 'read', 'list'];

export function contextToolCountKey(toolName: unknown): ContextToolCountKey | null {
    const name = normalizeContextToolName(toolName);
    if (name === 'read') return 'read';
    if (name === 'glob' || name === 'grep') return 'search';
    if (name === 'list') return 'list';
    return null;
}

export function summarizeContextTools(toolNames: readonly unknown[]): ContextToolCounts {
    const counts: ContextToolCounts = { read: 0, search: 0, list: 0 };
    for (const toolName of toolNames) {
        const key = contextToolCountKey(toolName);
        if (key) counts[key] += 1;
    }
    return counts;
}

export function isContextToolActive(part: unknown): boolean {
    return isToolPartActive(part);
}

/** 思考轨迹仍算探索过程；正文 / 非 context 工具才结束「探索中」。 */
export function isContextExploreSuccessorPart(input: {
    kind?: unknown;
    type?: unknown;
    toolName?: unknown;
}): boolean {
    if (input.kind === 'reasoning' || input.type === 'reasoning') {
        return false;
    }
    if (input.kind === 'justification' || input.type === 'text') {
        return true;
    }
    if (input.kind === 'tool' || input.type === 'tool') {
        return !isContextGroupTool(input.toolName);
    }
    return input.kind != null || input.type != null;
}

export function hasContextExploreSuccessor<T>(
    items: readonly T[],
    start: number,
    read: (item: T) => { kind?: unknown; type?: unknown; toolName?: unknown },
): boolean {
    for (let index = start; index < items.length; index += 1) {
        if (isContextExploreSuccessorPart(read(items[index]))) {
            return true;
        }
    }
    return false;
}

/**
 * 组内仍有未结算调用，或本轮还在进行且后面还没出现其他类型内容，都保持探索中。
 * 只看组内 tool status 会在批次空档（思考 / 等待下一轮探索工具）误判成「已探索」。
 *
 * 兼容旧调用：仅传 parts 数组时，退化为「组内是否有 active 工具」。
 */
export function isContextGroupExploring(
    partsOrInput:
        | readonly unknown[]
        | {
            parts: readonly unknown[];
            hasFollowingOtherType: boolean;
            isTurnLive: boolean;
        },
): boolean {
    if (!('parts' in partsOrInput)) {
        return partsOrInput.some((part) => isContextToolActive(part));
    }
    if (partsOrInput.parts.some((part) => isContextToolActive(part))) {
        return true;
    }
    if (partsOrInput.hasFollowingOtherType) {
        return false;
    }
    return partsOrInput.isTurnLive;
}

export function collectConsecutiveContextTools<T>(
    items: readonly T[],
    start: number,
    getToolName: (item: T) => unknown,
): { items: T[]; end: number } {
    const grouped: T[] = [];
    let index = start;
    while (index < items.length && isContextGroupTool(getToolName(items[index]))) {
        grouped.push(items[index]);
        index += 1;
    }
    return { items: grouped, end: index };
}

File: apps/mobile_flutter/lib/data/context_tool_grouping.dart
/// Port of `packages/ui/src/components/chat/message/parts/contextToolGrouping.ts`
/// and `toolRenderUtils.ts` context-tool names. Do not invent extra tools.
library;

import 'chat_timeline.dart';

const contextGroupToolNames = {'read', 'glob', 'grep', 'list'};

const _activeToolStatuses = {'pending', 'started', 'running'};

const _settledToolStatuses = {
  'completed',
  'error',
  'failed',
  'aborted',
  'timeout',
  'cancelled',
};

enum ContextToolCountKey { read, search, list }

const contextToolCountOrder = [
  ContextToolCountKey.search,
  ContextToolCountKey.read,
  ContextToolCountKey.list,
];

class ContextToolCounts {
  const ContextToolCounts({this.read = 0, this.search = 0, this.list = 0});

  final int read;
  final int search;
  final int list;

  int operator [](ContextToolCountKey key) {
    switch (key) {
      case ContextToolCountKey.read:
        return read;
      case ContextToolCountKey.search:
        return search;
      case ContextToolCountKey.list:
        return list;
    }
  }
}

class ConsecutiveContextTools {
  const ConsecutiveContextTools({required this.items, required this.end});

  final List<ChatPart> items;
  final int end;
}

String normalizeContextToolName(String? toolName) {
  final trimmed = toolName?.trim().toLowerCase() ?? '';
  if (trimmed.isEmpty) return '';
  final withoutIndex = trimmed.replaceFirst(RegExp(r':\d+$'), '');
  if (!withoutIndex.contains('.')) return withoutIndex;
  final parts = withoutIndex.split('.').where((part) => part.isNotEmpty).toList();
  return parts.isEmpty ? withoutIndex : parts.last;
}

bool isContextGroupTool(String? toolName) =>
    contextGroupToolNames.contains(normalizeContextToolName(toolName));

bool isToolPartActive(ChatPart part) {
  final status = (part.status ?? '').trim().toLowerCase();
  if (_activeToolStatuses.contains(status)) return true;
  if (_settledToolStatuses.contains(status)) return false;
  return false;
}

bool isContextExploreSuccessorPart(ChatPart part) {
  if (part.kind == ChatPartKind.text ||
      part.kind == ChatPartKind.mermaid ||
      part.kind == ChatPartKind.reasoning) {
    return true;
  }
  if (part.kind == ChatPartKind.permission) return false;
  if (part.kind == ChatPartKind.tool ||
      part.kind == ChatPartKind.fileOp ||
      part.kind == ChatPartKind.diff ||
      part.kind == ChatPartKind.task) {
    return !isContextGroupTool(part.toolName);
  }
  return true;
}

bool hasContextExploreSuccessor(List<ChatPart> items, int start) {
  for (var index = start; index < items.length; index += 1) {
    if (isContextExploreSuccessorPart(items[index])) return true;
  }
  return false;
}

bool isContextGroupExploring({
  required List<ChatPart> parts,
  required bool hasFollowingOtherType,
  required bool isTurnLive,
}) {
  if (parts.any(isToolPartActive)) return true;
  if (hasFollowingOtherType) return false;
  return isTurnLive;
}

ContextToolCountKey? contextToolCountKey(String? toolName) {
  final name = normalizeContextToolName(toolName);
  if (name == 'read') return ContextToolCountKey.read;
  if (name == 'glob' || name == 'grep') return ContextToolCountKey.search;
  if (name == 'list') return ContextToolCountKey.list;
  return null;
}

ContextToolCounts summarizeContextTools(Iterable<String?> toolNames) {
  var read = 0;
  var search = 0;
  var list = 0;
  for (final toolName in toolNames) {
    switch (contextToolCountKey(toolName)) {
      case ContextToolCountKey.read:
        read += 1;
      case ContextToolCountKey.search:
        search += 1;
      case ContextToolCountKey.list:
        list += 1;
      case null:
        break;
    }
  }
  return ContextToolCounts(read: read, search: search, list: list);
}

ConsecutiveContextTools collectConsecutiveContextTools(List<ChatPart> items, int start) {
  final grouped = <ChatPart>[];
  var index = start;
  while (index < items.length && isContextGroupTool(items[index].toolName)) {
    grouped.add(items[index]);
    index += 1;
  }
  return ConsecutiveContextTools(items: grouped, end: index);
}

bool messageHasRunningTool(ChatMessage message) => message.parts.any(isToolPartActive);

bool messageHasConfirmedFinalBody(ChatMessage message) {
  final text = message.parts
      .where((part) => part.kind == ChatPartKind.text)
      .map((part) => part.body?.trim() ?? '')
      .where((text) => text.isNotEmpty)
      .join();
  return text.isNotEmpty && !messageHasRunningTool(message);
}
